### PR TITLE
Allow Ray Tune callbacks to be passed into hyperopt and log model config

### PR DESCRIPTION
**Pull request by jeffkinnison**
This update makes three changes to hyperopt in Ludwig:

1.  Adds the `tune_callbacks` parameter to hyperopt to allow passing [Ray Tune Callbacks](https://docs.ray.io/en/latest/tune/api_docs/internals.html#callbacks) into Ray Tune
2. Updates the logic of the RayTuneExecutor to pass the callbacks into Ray Tune
3. Logs user-supplied and rendered model configs to trial logdirs

These changes support error surfacing, as well as #2343 and #2344

### CODE REVIEW

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [inline ludwig/hyperopt/execution.py:817]: Should this be false by default?

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [inline ludwig/hyperopt/execution.py:459]: @jeffkinnison We actually write this during `model.train()` here: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/api.py#L547

Is the reason to do this to get `model_hyperparameters.json` before we even get to `model.train()`? This would be useful is there's failures that may happen during preprocessing time since we only write `model_hyperparameters.json` after preprocessing.

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [inline ludwig/hyperopt/execution.py:453]: @jeffkinnison Would it make sense to potentially rename this to `trial_hyperparameters.json`?

Comment (arnavgarg1) [reply ludwig/hyperopt/execution.py:453]: Also, would it make more sense to do this with `modified_config` (pre-merge with defaults, post merge with sampled values) as supposed to just `hyperopt_dict["config"]`? I'm assuming we also want to capture the values that were actually sampled for this trial, which I believe is stored in config and then merged into `hyperopt_dict["config"]` in the substitute_parameters call below. Let me know if I'm missing something!

File: ludwig/hyperopt/run.py
Comment (arnavgarg1) [inline ludwig/hyperopt/run.py:10]: @jeffkinnison We'll want to move this into the try-except block on line 46, and then set `TuneCallback` to None or something like that. The reason is because Ray/Dask falls under "distributed" tests, so this will cause failures in the non-distributed tests when collecting all the tests to run

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [reply ludwig/hyperopt/execution.py:459]: Would also encourage using `MODEL_HYPERPARAMETERS_FILE_NAME` from `ludwig.globals` just incase this file name changes :)

**Comment by arnavgarg1** [comment]
Are we able to modify an existing test and pass in a very simple TuneCallback to make sure that tune_callbacks works as intended? Any of the ones in `test_hyperopt.py` or `test_hyperopt_ray.py` might be good candidates!

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py:459]: Yeah, I see that we make some modifications to the config at the start of training. It probably makes more sense to read the rendered config on training start.

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py:453]: Sure, I can do that.

**Comment by jeffkinnison** [comment]
> Are we able to modify an existing test and pass in a very simple TuneCallback to make sure that tune_callbacks works as intended? Any of the ones in `test_hyperopt.py` or `test_hyperopt_ray.py` might be good candidates!

I can do that for both.

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py:453]: That was a good call, and an oversight on my part, thanks!

File: ludwig/hyperopt/execution.py
Comment (ShreyaR) [inline ludwig/hyperopt/execution.py:813]: nit: Why not just add this as an optional argument to the function? That was, you can also add documentation for it in the docstring that explains why it's needed.

**Review by ShreyaR — APPROVED** [review]
lgtm! Feel free to merge once comment is addressed and tests are passing.

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py:813]: This is an artifact of an earlier build I'm going to remove it now.

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py]: This should be addressed by [7b40418](https://github.com/ludwig-ai/ludwig/pull/2640/commits/7b40418caeba34b8ab3c914984282f07542a75cf)

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [inline ludwig/hyperopt/execution.py:452]: nit: Can we update this comment to 

```Write out the unmerged config with sampled parameters to the trial's local directory```

File: ludwig/hyperopt/execution.py
Comment (arnavgarg1) [reply ludwig/hyperopt/execution.py:813]: @jeffkinnison Was this supposed to be updated based on your response to Shreya's comment? Seeing the same thing as before - let me know!

File: ludwig/hyperopt/run.py
Comment (arnavgarg1) [reply ludwig/hyperopt/run.py:10]: @jeffkinnison I think you'll still need to address this to prevent the test failures in the non-distributed tests

File: tests/integration_tests/test_hyperopt_ray.py
Comment (arnavgarg1) [inline tests/integration_tests/test_hyperopt_ray.py:143]: I think this should be 

```
super().on_trial_complete(iteration, trials, trial, **info)
```

**Review by arnavgarg1 — CHANGES_REQUESTED** [review]
This is almost there. I think there's a couple of small changes

File: tests/integration_tests/test_hyperopt_ray.py
Comment (jeffkinnison) [reply tests/integration_tests/test_hyperopt_ray.py:363]: Ah, somehow that was left out of the last commit.

File: ludwig/hyperopt/run.py
Comment (jeffkinnison) [reply ludwig/hyperopt/run.py]: I removed it entirely since it was only used for type hints. Replaced with a string in [a073cc0](https://github.com/ludwig-ai/ludwig/pull/2640/commits/a073cc00aec3e3915c4cdd143aabbee7a247c816)

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py]: Removed in [a073cc0](https://github.com/ludwig-ai/ludwig/pull/2640/commits/a073cc00aec3e3915c4cdd143aabbee7a247c816)

File: ludwig/hyperopt/execution.py
Comment (jeffkinnison) [reply ludwig/hyperopt/execution.py]: Updated in [a073cc0](https://github.com/ludwig-ai/ludwig/pull/2640/commits/a073cc00aec3e3915c4cdd143aabbee7a247c816)

File: tests/integration_tests/test_hyperopt_ray.py
Comment (jeffkinnison) [reply tests/integration_tests/test_hyperopt_ray.py]: The code now exists in [a073cc0](https://github.com/ludwig-ai/ludwig/pull/2640/commits/a073cc00aec3e3915c4cdd143aabbee7a247c816)

File: tests/integration_tests/test_hyperopt_ray.py
Comment (jeffkinnison) [reply tests/integration_tests/test_hyperopt_ray.py]: Fixed in [a073cc0](https://github.com/ludwig-ai/ludwig/pull/2640/commits/a073cc00aec3e3915c4cdd143aabbee7a247c816)

File: tests/integration_tests/test_hyperopt_ray.py
Comment (arnavgarg1) [inline tests/integration_tests/test_hyperopt_ray.py:26]: Accidental import?

File: tests/integration_tests/test_hyperopt_ray.py
Comment (arnavgarg1) [inline tests/integration_tests/test_hyperopt_ray.py:27]: Wondering if this import is needed and if it'll cause the same error that the callback was throwing - would be good to double check!

Comment (jeffkinnison) [reply tests/integration_tests/test_hyperopt_ray.py:27]: Pre-commit CI bot added that too.

**Review by arnavgarg1 — APPROVED** [review]
Looks great, thanks @jeffkinnison ! Would just resolve merge conflicts and make sure the tests pass before merging!

File: tests/integration_tests/test_hyperopt_ray.py
Comment (arnavgarg1) [inline tests/integration_tests/test_hyperopt_ray.py:156]: Is it worth adding a check here to make sure the loaded json isn't empty?

Comment (arnavgarg1) [reply tests/integration_tests/test_hyperopt_ray.py:156]: Same for the json load below

Comment (jeffkinnison) [reply tests/integration_tests/test_hyperopt_ray.py:156]: Sure, I can add that.

File: tests/integration_tests/test_hyperopt_ray.py
Comment (arnavgarg1) [inline tests/integration_tests/test_hyperopt_ray.py:144]: This might just need a # noqa

**Pull request merged**
